frog/responsive-images.rkt

```diff
@@ -0,0 +1,117 @@
+
+(define/contract (resize-image in new-width out)
+  (path? number? path? . -> . boolean?)
+  ;; Depend on ImageMagick
+  (define (magick-resize in w out)
+    (system (format "magick \"~a\" -geometry ~a \"~a\"" (path->string in) w (path->string out))))
+  (define resize magick-resize)
+  (prn1 "Resizing image ~a to ~a pixels... " (path->string in) new-width)
+  (if (> (current-verbosity) 1)
+      (time (resize in new-width out))
+      (resize in new-width out)))
+
+(module+ test
+  (when magick-available?
+    (define out (string->path "/tmp/resized.gif"))
+    (test-eq? "resize" (begin
+                         (resize-image (string->path "../example/img/480px-image.gif") 10 out)
+                         (image-width out))
+              10)
+    (delete-file out))) ; catch exn?
+
+(define/contract (get-images image-path)
+  (path? . -> . (values pair? (listof pair?)))
+  (define output-path (string->path (current-image-output-dir)))
+  (unless (directory-exists? output-path)
+    (make-directory output-path))
+  (let* ([orig-size (image-width image-path)]
+         [sizes (filter ((curry >=) orig-size) (current-image-sizes))])
+    (values (cons image-path orig-size)
+            (for/list ([width sizes])
+              (let* ([filename (file-name-from-path image-path)]
+                     [extension (get-extension filename)]
+                     [out-path
+                      (build-path
+                       output-path
+                       (replace-extension
+                        filename
+                        (string-append "-" (number->string width) "." extension)))])
+                (unless (and (file-exists? out-path)
+                             (< (file-or-directory-modify-seconds image-path)
+                                (file-or-directory-modify-seconds out-path)))
+                  (resize-image image-path width out-path))
+                (cons out-path width))))))
+
+(define default-image-idx
+  (list-index ((curry =) (current-image-default-size)) (current-image-sizes)))
+
+(define/contract (make-responsive url)
+  (url? . -> . (listof pair?))
+  (define (make-url-string path)
+    (url->string (combine-url/relative
+                  (string->url "/")
+                  (path->string path))))
+  (define (absolute->relative s) (string->path (substring (path->string s) 1)))
+  (let ([image-path (absolute->relative (url->path url))])
+    (let-values ([(orig srcset) (get-images image-path)])
+      (let ([src (make-url-string (car (if (> (length srcset) default-image-idx)
+                                           (list-ref srcset default-image-idx)
+                                           orig)))]
+            [srcset-string
+             (string-join (map (λ (srcdef)
+                                 (format "~a ~aw" (make-url-string (car srcdef)) (cdr srcdef)))
+                               srcset)
+                          ", ")])
+        `([src ,src]
```

Comment:
This is very much an optional bit of feedback. Feel free to disregard.

The Racket style guide says to prefer `define` over `let` -- mainly because it avoids excessive nesting/indentation. Although I generally agree, I'm not adamant. I also write a lot of Emacs Lisp and Clojure, where it's all `let`. Shrug.

These lines with `let` inside `let-values` inside `let`? Might be an example where using `define` (and `define-values`) would be nicer?  Something like:

```rkt
(define image-path (absolute->relative (url->path url))
(define-values (orig srcset) (get-images image-path))
(define src ....)
(define srcset-string ....)
```

Again, just something to consider. Not a substantive issue, and I don't want to be the Style Police. :)  I'd happily merge this as-is.